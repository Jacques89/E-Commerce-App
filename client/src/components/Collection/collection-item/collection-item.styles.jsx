import styled from 'styled-components'
import CustomButton from '../../Button/custom-button/custom-button.component'

export const CollectionItemDiv = styled.div`
    width: 22vw;
    display: flex;
    flex-direction: column;
    height: 21.875rem;
    align-items: center;
    position: relative;
    margin-bottom: 1.25rem;
    cursor: pointer;

    &:hover {
        .image {
            opacity: 0.8;
        }
        button {
            opacity: 0.85;
            display: flex;
        }
    }

    @media screen and (max-width: 800px) {
        width: 40vw;

        &:hover {
            .image {
                opacity: unset;
            }
            button {
                opacity: unset;
            }
        }
    }

    @media screen and (max-width: 554px) {
        width: 40vw;
        display: flex;
        &:hover {
            .image {
                opacity: unset;
            }
            button {
                opacity: unset;
            }
        }
    }
`
CollectionItemDiv.displayName = 'CollectionItemStyles'

export const AddButton = styled(CustomButton)`
    width: 80%;
    opacity: 0.7;
    position: absolute;
    top: 15.938rem;
    display: none;

    @media screen and (max-width: 800px) {
        display: block;
        opacity: 0.9;
        min-width: unset;
        padding: 0 0.625rem;
    }
`
AddButton.displayName = 'AddButtonStyles'

export const BackgroundImage = styled.div`
    width: 100%;
    height: 95%;
    background-size: cover;
    background-position: center;
    margin-bottom: 0.313rem;
    background-image: ${({ imageUrl }) => `url(${imageUrl})`};

    @media screen and (max-width: 800px) {
        width: 77vw;
        background-size: 41rem;
        background-repeat: no-repeat;
    }
`
BackgroundImage.displayName = 'BackgroundImageStyles'

export const CollectionFooterDiv = styled.div`
    width: 100%;
    height: 5%;
    display: flex;
    justify-content: space-between;
    font-size: 1.125rem;

    @media screen and (max-width: 800px) {
        width: 77vw;
    }
`
CollectionFooterDiv.displayName = 'CollectionFooterStyles'

export const NameSpan = styled.span`
    width: 90%;
    margin-bottom: 0.938rem;
`
NameSpan.displayName = 'NameSpanStyles'

export const PriceSpan = styled.span`
    width: 10%;
    text-align: right;
`
PriceSpan.displayName = 'PriceSpanStyles'
