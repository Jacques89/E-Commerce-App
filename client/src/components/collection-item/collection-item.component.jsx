import React from 'react'
import { withRouter } from 'react-router-dom'
import PropTypes from 'prop-types'

import {
    CollectionItemDiv,
    CollectionFooterDiv,
    AddButton,
    BackgroundImage,
    NameSpan,
    PriceSpan
} from './collection-item.styles'

export const CollectionItem = ({ item, addItem, history, match, routeName }) => {
    const { id, name, price, imageUrl } = item
    

    return (
        <CollectionItemDiv>
            <BackgroundImage
                className='image'
                imageUrl={imageUrl}
                onClick={() => {
                    if (match.url === '/shop') {
                        history.push(`${match.url}/${routeName}/${id}`)
                    } else if (Number.isNaN(parseInt(match.url.slice(-1)))) {
                        history.push(`${match.url}/${id}`)
                    } else {
                        history.push(`/shop/${routeName}/${id}`)
                    }
                }}
            />
            <CollectionFooterDiv>
                <NameSpan>{name}</NameSpan>
                <PriceSpan>{price}â‚¬</PriceSpan>
            </CollectionFooterDiv>
            <AddButton onClick={() => addItem(item)} inverted>
                Add to cart
            </AddButton>
        </CollectionItemDiv>
    )
}

export default withRouter(CollectionItem)

CollectionItem.propTypes = {
    item: PropTypes.object.isRequired,
    addItem: PropTypes.func.isRequired,
    history: PropTypes.object,
    match: PropTypes.object,
    routeName: PropTypes.string
}