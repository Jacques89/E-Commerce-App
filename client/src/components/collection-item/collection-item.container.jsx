import { connect } from 'react-redux'
import { addItem } from '../../redux/cart/cart.actions'
import CollectionItem from './collection-item.component'

const mapDispatchToProps = (dispatch) => ({
    addItem: (item) => dispatch(addItem(item))
})

const CollectionItemContainer = connect(null, mapDispatchToProps)(CollectionItem)

export default CollectionItemContainer
