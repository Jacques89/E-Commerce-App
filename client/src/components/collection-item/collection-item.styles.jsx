import styled from 'styled-components'
import CustomButton from '../custom-button/custom-button.component'

export const CollectionItemContainer = styled.div`
    width: 22vw;
    display: flex;
    flex-direction: column;
    height: 21.875rem;
    align-items: center;
    position: relative;

    &:hover {
        .image {
            opacity: 0.8;
        }

        button {
            opacity: 0.85;
            display: flex;
        }
    }

    @media screen and (max-width: 800px) {
        width: 40vw;
        grid-template-columns: 1fr;
        display: flex;
        justify-content: space-between;

        &:hover {
            .image {
                opacity: unset;
            }

            button {
                opacity: unset;
            }
        }
    }
`

export const AddButton = styled(CustomButton)`
    width: 80%;
    opacity: 0.7;
    position: absolute;
    top: 15.938rem;
    display: none;

    @media screen and (max-width: 800px) {
        display: block;
        opacity: 0.9;
        min-width: unset;
        top: 13.688rem;
        padding: 0 0.625rem 0 0.625rem;
    }
`

AddButton.displayName = 'AddButton'

export const BackgroundImage = styled.div`
    width: 100%;
    height: 95%;
    background-size: cover;
    background-position: center;
    margin-bottom: 0.313rem;
    background-image: ${({ imageUrl }) => `url(${imageUrl})`};

    @media screen and (max-width: 800px) {
        width: 80vw;
        height: 95%;
    }
`

BackgroundImage.displayName = 'BackgroundImage'

export const CollectionFooterContainer = styled.div`
    width: 100%;
    height: 5%;
    display: flex;
    justify-content: space-between;
    font-size: 1.4rem;

    @media screen and (max-width: 800px) {
        width: 80vw;
        height: 20%;
        font-size: 2.25rem;
    }
`

CollectionFooterContainer.displayName = 'CollectionFooterContainer'

export const NameContainer = styled.span`
    width: 90%;
    margin-bottom: 0.938rem;
`

NameContainer.displayName = 'NameContainer'

export const PriceContainer = styled.span`
    text-align: right;
`

PriceContainer.displayName = 'PriceContainer'
