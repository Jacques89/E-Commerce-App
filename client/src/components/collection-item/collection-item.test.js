import React from 'react'
import { mount } from 'enzyme'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'

import { BrowserRouter as Router } from 'react-router-dom'

import { CollectionItem } from './collection-item.component'

describe('CollectionItem component', () => {
    let wrapper
    let mockAddItem
    const imageUrl = 'www.testImage.com'
    const mockName = 'blackCDJ'
    const mockPrice = 10

    const mockStore = configureMockStore()
    const store = mockStore({})

    beforeEach(() => {
        mockAddItem = jest.fn()

        const mockProps = {
            item: {
                imageUrl: imageUrl,
                price: mockPrice,
                name: mockName
            },
            addItem: mockAddItem
        }

        wrapper = mount(
            <Provider store={store}>
                <Router>
                    <CollectionItem item={mockProps.item} addItem={mockProps.addItem} />
                </Router>
            </Provider>
        )
    })

    it('should render CollectionItem component', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('should call addItem when AddButton clicked', () => {
        wrapper.find('AddButtonStyles').simulate('click')
        expect(mockAddItem).toHaveBeenCalled()
    })

    it('should render imageUrl as a prop on BackgroundImage', () => {
        expect(wrapper.find('BackgroundImageStyles').prop('imageUrl')).toBe(imageUrl)
    })

    it('should render name prop in NameSpan', () => {
        expect(wrapper.find('NameSpanStyles').text()).toBe(mockName)
    })

    it('should render price prop in PriceSpan', () => {
        const price = parseInt(wrapper.find('PriceSpanStyles').text())
        expect(price).toBe(mockPrice)
    })
})
