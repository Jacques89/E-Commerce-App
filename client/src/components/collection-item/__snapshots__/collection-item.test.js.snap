// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionItem component should render CollectionItem component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <CollectionItem
        addItem={[MockFunction]}
        item={
          Object {
            "imageUrl": "www.testImage.com",
            "name": "blackCDJ",
            "price": 10,
          }
        }
      >
        <CollectionItemStyles>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bwzfXH",
                  "isStatic": false,
                  "lastClassName": "dvtpVT",
                  "rules": Array [
                    "
    width: 22vw;
    display: flex;
    flex-direction: column;
    height: 21.875rem;
    align-items: center;
    position: relative;
    margin-bottom: 1.25rem;
    cursor: pointer;

    &:hover {
        .image {
            opacity: 0.8;
        }
        button {
            opacity: 0.85;
            display: flex;
        }
    }

    @media screen and (max-width: 800px) {
        width: 40vw;

        &:hover {
            .image {
                opacity: unset;
            }
            button {
                opacity: unset;
            }
        }
    }

    @media screen and (max-width: 554px) {
        width: 40vw;
        display: flex;
        &:hover {
            .image {
                opacity: unset;
            }
            button {
                opacity: unset;
            }
        }
    }
",
                  ],
                },
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bwzfXH",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="sc-bwzfXH dvtpVT"
            >
              <BackgroundImageStyles
                className="image"
                imageUrl="www.testImage.com"
                onClick={[Function]}
              >
                <StyledComponent
                  className="image"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-bxivhb",
                        "isStatic": false,
                        "lastClassName": "hFVDgt",
                        "rules": Array [
                          "
    width: 100%;
    height: 95%;
    background-size: cover;
    background-position: center;
    margin-bottom: 0.313rem;
    background-image: ",
                          [Function],
                          ";

    @media screen and (max-width: 554px) {
        width: 200%
    }
",
                        ],
                      },
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-bxivhb",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  imageUrl="www.testImage.com"
                  onClick={[Function]}
                >
                  <div
                    className="sc-bxivhb hFVDgt image"
                    onClick={[Function]}
                  />
                </StyledComponent>
              </BackgroundImageStyles>
              <CollectionFooterStyles>
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-ifAKCX",
                        "isStatic": false,
                        "lastClassName": "VEwmV",
                        "rules": Array [
                          "
    width: 100%;
    height: 5%;
    display: flex;
    justify-content: space-between;
    font-size: 1.125rem;

    @media screen and (max-width: 554px) {
        width: 200%
    }
",
                        ],
                      },
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-ifAKCX",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <div
                    className="sc-ifAKCX VEwmV"
                  >
                    <NameSpanStyles>
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-EHOje",
                              "isStatic": false,
                              "lastClassName": "kXpvxx",
                              "rules": Array [
                                "
    width: 90%;
    margin-bottom: 0.938rem;
",
                              ],
                            },
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-EHOje",
                            "target": "span",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                      >
                        <span
                          className="sc-EHOje kXpvxx"
                        >
                          blackCDJ
                        </span>
                      </StyledComponent>
                    </NameSpanStyles>
                    <PriceSpanStyles>
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-bZQynM",
                              "isStatic": false,
                              "lastClassName": "jyrdPw",
                              "rules": Array [
                                "
    width: 10%;
    text-align: right;
",
                              ],
                            },
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-bZQynM",
                            "target": "span",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                      >
                        <span
                          className="sc-bZQynM jyrdPw"
                        >
                          10
                          â‚¬
                        </span>
                      </StyledComponent>
                    </PriceSpanStyles>
                  </div>
                </StyledComponent>
              </CollectionFooterStyles>
              <AddButtonStyles
                inverted={true}
                onClick={[Function]}
              >
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-htpNat",
                        "isStatic": false,
                        "lastClassName": "ixxrhM",
                        "rules": Array [
                          "
    width: 80%;
    opacity: 0.7;
    position: absolute;
    top: 15.938rem;
    display: none;

    @media screen and (max-width: 800px) {
        display: block;
        opacity: 0.9;
        min-width: unset;
        padding: 0 0.625rem;
    }
",
                        ],
                      },
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-htpNat",
                      "target": [Function],
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  inverted={true}
                  onClick={[Function]}
                >
                  <CustomButton
                    className="sc-htpNat ixxrhM"
                    inverted={true}
                    onClick={[Function]}
                  >
                    <styled.button
                      className="sc-htpNat ixxrhM"
                      inverted={true}
                      onClick={[Function]}
                    >
                      <StyledComponent
                        className="sc-htpNat ixxrhM"
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "sc-bdVaJa",
                              "isStatic": false,
                              "lastClassName": "egadog",
                              "rules": Array [
                                "
    min-width: 10.313rem;
    width: auto;
    height: 3.125rem;
    letter-spacing: 0.031rem;
    line-height: 3.125rem;
    padding: 0 2.188rem 0 2.188rem;
    font-size: 0.938rem;
    text-transform: uppercase;
    font-family: 'Open Sans Condensed';
    font-weight: bolder;
    cursor: pointer;
    display: flex;
    justify-content: center;

    ",
                                [Function],
                                "
",
                              ],
                            },
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "sc-bdVaJa",
                            "target": "button",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                        inverted={true}
                        onClick={[Function]}
                      >
                        <button
                          className="sc-bdVaJa egadog sc-htpNat ixxrhM"
                          onClick={[Function]}
                        >
                          Add to cart
                        </button>
                      </StyledComponent>
                    </styled.button>
                  </CustomButton>
                </StyledComponent>
              </AddButtonStyles>
            </div>
          </StyledComponent>
        </CollectionItemStyles>
      </CollectionItem>
    </Router>
  </BrowserRouter>
</Provider>
`;
